instrumentation/opentelemetry-instrumentation-urllib/src/opentelemetry/instrumentation/urllib/__init__.py

```diff
@@ -0,0 +1,233 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""
+This library allows tracing HTTP requests made by the
+`urllib https://docs.python.org/3/library/urllib.html>`_ library.
+
+Usage
+-----
+
+.. code-block:: python
+
+    from urllib import request
+    from libs.opentelemetry.instrumentation.urllib import URLLibInstrumentor
+
+    # You can optionally pass a custom TracerProvider to
+    URLLibInstrumentor().instrument()
+
+    req = request.Request('https://postman-echo.com/post', method="POST")
+    r = request.urlopen(req)
+
+API
+---
+"""
+
+import functools
+import types
+from urllib.request import (  # pylint: disable=no-name-in-module,import-error
+    OpenerDirector,
+    Request,
+)
+
+from opentelemetry import context, propagators
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.metric import (
+    HTTPMetricRecorder,
+    HTTPMetricType,
+    MetricMixin,
+)
+from opentelemetry.instrumentation.urllib.version import __version__
+from opentelemetry.instrumentation.utils import http_status_to_status_code
+from opentelemetry.trace import SpanKind, get_tracer
+from opentelemetry.trace.status import Status, StatusCode
+
+# A key to a context variable to avoid creating duplicate spans when instrumenting
+_SUPPRESS_URLLIB_INSTRUMENTATION_KEY = "suppress_urllib_instrumentation"
+
+
+# pylint: disable=unused-argument
+# pylint: disable=R0915
+def get_default_span_name(method):
+    """Default implementation for name_callback, returns HTTP {method_name}."""
+    return "HTTP {}".format(method).strip()
+
+
+def _instrument(tracer_provider=None, span_callback=None, name_callback=None):
+    """Enables tracing of all requests calls that go through
+    :code:`urllib.Client._make_request`"""
+
+    opener_open = OpenerDirector.open
+
+    @functools.wraps(opener_open)
+    def instrumented_request(opener, fullurl, data=None, timeout=None):
```

Comment:
it might make sense to allow *args / *kwargs here, since new methods may be added that need to be passed through to open. 

Does that make sense here?

Reply:
Not sure, as AFAIK `open` does not support specific arguments `open(self, fullurl, data=None, timeout=socket._GLOBAL_DEFAULT_TIMEOUT)`. I'd preferred left it as is. We can always add `*args`, `**kwargs` by request.

Reply:
The issue with that approach is that you're fixing an issue for people once it breaks them. I'm a bit wary of inconveniencing users knowingly. 

That all said, this will probably only change over major version of Python, so the likelyhood is lower.